README.md
# alerty

A CLI utility to send an alert if a given command failed. 

## How Useful?

I use `alerty` to run commands in cron to send alerts if cron commands fail.

```
0 * * * * alerty -c /etc/alerty/alerty.yml -- /path/to/script --foo FOO --bar
```

## Installation

```
gem install alerty
```

## Configuration

You can write a configuration file located at `/etc/alerty/alerty.yml` (You can configure this path by `ALERTY_CONFIG_FILE` environment variable, or `-c` option):

```
log_path: /var/tmp/alerty.log
log_level: 'info'
log_shift_age: 10
log_shift_size: 10485760
timeout: 10
lock_path: /tmp/lock
retry_limit: 2
retry_wait: 10
plugins:
  - type: stdout
```

[example.yml](./example.yml)

### CLI Example

```
$ alerty -c example.yml -- ls -l /something_not_exist
```

### CLI Help

```
Usage: alerty [options] -- command
    -c, --config CONFIG_FILE         config file path (default: /etc/alerty/alerty.yml)
        --log LOG_FILE               log file path (default: STDOUT)
    -l, --log-level LOG_LEVEL        log level (default: warn)
        --log-shift-age SHIFT_AGE    Number of old log files to keep (default: 0 which means no log rotation)
        --log-shift-size SHIFT_SIZE  Maximum logfile size in bytes (default: 1048576)
    -t, --timeout SECONDS            timeout the command (default: no timeout)
        --lock LOCK_FILE             exclusive lock file to prevent running a command duplicatedly (default: no lock)
        --retry-limit NUMBER         number of retries (default: 0)
        --retry-wait SECONDS         retry interval = retry wait +/- 12.5% randomness (default: 1.0)
    -d, --debug                      debug mode
        --dotenv                     Load environment variables from .env file with dotenv
```

### Experimental: Send alert from STDIN

This interface allows us to send notification even if a command does not fail.

CLI Example:

```
$ echo 'this is a test' | alerty -c example.yml
```

## Plugins

Following plugins are available:

* [stdout](./lib/alerty/plugin/stdout.rb)
* [file](./lib/alerty/plugin/file.rb)
* [exec](./lib/alerty/plugin/exec.rb)
* [sonots/alerty-plugin-ikachan](https://github.com/sonots/alerty-plugin-ikachan)
* [sonots/alerty-plugin-amazon_sns](https://github.com/sonots/alerty-plugin-amazon_sns)
* [sonots/alerty-plugin-slack](https://github.com/sonots/alerty-plugin-slack)
* [civitaspo/alerty-plugin-mail](https://github.com/civitaspo/alerty-plugin-mail)
* [inokappa/alerty-plugin-datadog_event](https://github.com/inokappa/alerty-plugin-datadog_event)
* [inokappa/alerty-plugin-amazon_cloudwatch_logs](https://github.com/inokappa/alerty-plugin-amazon_cloudwatch_logs)
* [inokappa/alerty-plugin-post_im_kayac](https://github.com/inokappa/alerty-plugin-post_im_kayac)

## Plugin Architecture

### Naming Convention

You must follow following naming conventions:

* gem name: alerty-plugin-xxx (xxx_yyy)
* file name: lib/alerty/plugin/xxx.rb (xxx_yyy.rb)
* class name: Alerty::Plugin:Xxx (XxxYyy)

### Interface
u
What you have to implement is `#initialize` and `#alert` methods. Here is an example of `file` plugin:

```ruby
require 'json'

class Alerty
  class Plugin
    class File
      def initialize(config)
        raise ConfigError.new('file: path is not configured') unless config.path
        @path = config.path
      end

      def alert(record)
        ::File.open(@path, 'a') do |io|
          io.puts record.to_json
        end
      end
    end
  end
end
```

### config

`config` is created from the configuration file: 

```
plugins:
  - type: foobar
    key1: val1
    key2: val2
```

`config.key1` and `config.key2` are availabe in the above config. 

### record

`record` is a hash whose keys are symbols of followings:

* hostname: hostname
* command: the executed command
* exitstatus: the exit status of the executed command
* output: the output of the exectued command
* started_at: the time when command executed in epoch time.
* duration: the duration which the command execution has taken in seconds. 
* retries: the number of retries

## ChangeLog

See [CHANGELOG.md](CHANGELOG.md) for details.

### Licenses

See [LICENSE](LICENSE)


CHANGELOG.md
# 0.4.0 (2017/09/29)

Enhancements:

* [experimental] send alert from stdin

# 0.3.0 (2017/03/20)

Enhancements:

* Use popen2e rather than `2>&1`

# 0.2.3 (2017/01/20)

Enhancements:

* Enrich debug print
* Add --version option

# 0.2.2 (2017/01/20)

Fixes:

* Fix log_shift_age and log_shift_size config were not working (instead, shift_age and shift_size were working)

# 0.2.1 (2016/09/25)

Enhancements:

* Add --dotenv option to load environment variables from .env file

# 0.2.0 (2016/09/25)

Enhancements:

* Allow erb in config yaml

# 0.1.1 (2016/09/05)

Enhancements:

* Bundler.with_clean_env

# 0.1.0 (2016/03/28)

Enhancements:

* Add --log-shift-age option
* Add --log-shift-size option

# 0.0.9 (2015/11/23)

Enhancements:

* Support retry

# 0.0.8 (2015/08/15)

Enhancements:

* Send alert if timeout or locked occurs

# 0.0.7 (2015/08/15)

Enhancements:

* Change -l option from --lock to --log-level
* Add -d (--debug) option, same with -l debug

# 0.0.6 (2015/08/14)

Enhancements:

* Add info log to show command result

# 0.0.5 (2015/08/14)

Changes:

* Change default conf path from /etc/sysconfig/alerty to /etc/alerty/alerty.yml

# 0.0.4 (2015/08/14)

Enchancements:

* Add exec plugin
* Pass `hostname` to plugins

# 0.0.3 (2015/08/14)

Enchancements:

* Pass more information to plugins
  * command
  * exitstatus
  * output
  * started_at
  * duration

# 0.0.2 (2015/08/13)

Enchancements:

* Allow to configure timeout and lock_path in config file (options outcome)
* Allow to configure log_path and log_level as command options (options outcome)

# 0.0.1 (2015/08/13)

first version


lib/alerty/cli.rb
require 'optparse'
require 'socket'
require_relative '../alerty'

class Alerty
  class CLI
    def parse_options(argv = ARGV)
      op = OptionParser.new
      op.banner += ' -- command'

      (class<<self;self;end).module_eval do
        define_method(:usage) do |msg|
          puts op.to_s
          puts "error: #{msg}" if msg
          exit 1
        end
      end

      opts = {}
      op.on('-c', '--config CONFIG_FILE', "config file path (default: /etc/alerty/alerty.yml)") {|v|
        opts[:config_path] = v
      }
      op.on('--log LOG_FILE', "log file path (default: STDOUT)") {|v|
        opts[:log_path] = v
      }
      op.on('-l', '--log-level LOG_LEVEL', "log level (default: warn)") {|v|
        opts[:log_level] = v
      }
      op.on('--log-shift-age SHIFT_AGE', "Number of old log files to keep (default: 0 which means no log rotation)") {|v|
        opts[:log_shift_age] = Integer(v)
      }
      op.on('--log-shift-size SHIFT_SIZE', "Maximum logfile size in bytes (default: 1048576)") {|v|
        opts[:log_shift_age] = Integer(v)
      }
      op.on('-t', '--timeout SECONDS', "timeout the command (default: no timeout)") {|v|
        opts[:timeout] = v.to_f
      }
      op.on('--lock LOCK_FILE', "exclusive lock file to prevent running a command duplicatedly (default: no lock)") {|v|
        opts[:lock_path] = v
      }
      op.on('--retry-limit NUMBER', "number of retries (default: 0)") {|v|
        opts[:retry_limit] = v.to_i
      }
      op.on('--retry-wait SECONDS', "retry interval = retry wait +/- 12.5% randomness (default: 1.0)") {|v|
        opts[:retry_wait] = v.to_f
      }
      op.on('--dotenv', "Load environment variables from .env file with dotenv") {|v|
        opts[:dotenv] = true
      }
      op.on('-d', '--debug', "debug mode") {|v|
        opts[:debug] = true
      }
      op.on('-v', '--version', "print version") {|v|
        puts Alerty::VERSION
        exit 0
      }

      op.parse!(argv)
      opts[:command] = argv.join(' ') || ''

      opts
    end

    def run
      begin
        opts = parse_options
      rescue OptionParser::InvalidOption => e
        usage e.message
      end
  
      Config.configure(opts)
      PluginFactory.plugins # load plugins in early stage

      if !opts[:command].empty?
        command = Command.new(command: opts[:command])
        record = command.run
        unless record[:exitstatus] == 0
          Alerty.send(record)
          exit record[:exitstatus]
        end
      else
        begin
          stdin = $stdin.read_nonblock(100 * 1024 * 1024)
          record = {hostname: Socket.gethostname, output: stdin}
          Alerty.send(record)
        rescue IO::EAGAINWaitReadable => e
          usage 'command argument or STDIN is required'
        end
      end
    end
  end
end
